File: rust/day24_part1_2023.rs

use std::fs::File;
use std::io::{self, BufRead};

#[derive(Debug, Clone, Copy)]
struct Hailstone {
    px: f64,
    py: f64,
    vx: f64,
    vy: f64,
}

fn parse_hailstone(line: &str) -> Option<Hailstone> {
    let parts: Vec<&str> = line.split(" @ ").collect();
    if parts.len() != 2 {
        return None;
    }

    let pos_parts: Vec<&str> = parts[0].split(", ").collect();
    let vel_parts: Vec<&str> = parts[1].split(", ").collect();

    if pos_parts.len() != 3 || vel_parts.len() != 3 {
        return None;
    }

    Some(Hailstone {
        px: pos_parts[0].trim().parse().ok()?,
        py: pos_parts[1].trim().parse().ok()?,
        vx: vel_parts[0].trim().parse().ok()?,
        vy: vel_parts[1].trim().parse().ok()?,
    })
}

fn intersection(h1: &Hailstone, h2: &Hailstone) -> Option<(f64, f64)> {
    let det = h1.vx * h2.vy - h2.vx * h1.vy;
    if det == 0.0 {
        return None; // Parallel lines
    }

    let t1 = ((h2.px - h1.px) * h2.vy - (h2.py - h1.py) * h2.vx) / det;
    let t2 = ((h2.px - h1.px) * h1.vy - (h2.py - h1.py) * h1.vx) / det;

    if t1 < 0.0 || t2 < 0.0 {
        return None; // Intersection in the past
    }

    let x = h1.px + t1 * h1.vx;
    let y = h1.py + t1 * h1.vy;
    Some((x, y))
}

fn main() -> io::Result<()> {
    let file = File::open("input.txt")?;
    let reader = io::BufReader::new(file);

    let hailstones: Vec<Hailstone> = reader
        .lines()
        .filter_map(|line| line.ok())
        .filter_map(|line| parse_hailstone(&line))
        .collect();

    let min_coord = 200000000000000.0;
    let max_coord = 400000000000000.0;

    let mut intersection_count = 0;
    for i in 0..hailstones.len() {
        for j in i + 1..hailstones.len() {
            if let Some((x, y)) = intersection(&hailstones[i], &hailstones[j]) {
                if x >= min_coord && x <= max_coord && y >= min_coord && y <= max_coord {
                    intersection_count += 1;
                }
            }
        }
    }

    println!("{}", intersection_count);

    Ok(())
}

File: java/day24_part1_2023.java

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class HailstonePaths {

    static class Hailstone {
        long px, py, vx, vy;

        public Hailstone(long px, long py, long vx, long vy) {
            this.px = px;
            this.py = py;
            this.vx = vx;
            this.vy = vy;
        }
    }

    public static void main(String[] args) {
        String inputFile = "input.txt";
        List<Hailstone> hailstones = new ArrayList<>();

        try (BufferedReader br = new BufferedReader(new FileReader(inputFile))) {
            String line;
            while ((line = br.readLine()) != null) {
                String[] parts = line.split(" @ ");
                String[] pos = parts[0].split(", ");
                String[] vel = parts[1].split(", ");
                hailstones.add(new Hailstone(
                        Long.parseLong(pos[0].trim()),
                        Long.parseLong(pos[1].trim()),
                        Long.parseLong(vel[0].trim()),
                        Long.parseLong(vel[1].trim())
                ));
            }
        } catch (IOException e) {
            System.err.println("Error reading file: " + e.getMessage());
            return;
        }

        long minCoord = 200000000000000L;
        long maxCoord = 400000000000000L;
        int intersections = 0;

        for (int i = 0; i < hailstones.size(); i++) {
            for (int j = i + 1; j < hailstones.size(); j++) {
                if (intersect(hailstones.get(i), hailstones.get(j), minCoord, maxCoord)) {
                    intersections++;
                }
            }
        }

        System.out.println(intersections);
    }

    static boolean intersect(Hailstone h1, Hailstone h2, long min, long max) {
        // Solve for t1 and t2:
        // h1.px + t1 * h1.vx = h2.px + t2 * h2.vx
        // h1.py + t1 * h1.vy = h2.py + t2 * h2.vy

        // Rearrange to:
        // t1 * h1.vx - t2 * h2.vx = h2.px - h1.px
        // t1 * h1.vy - t2 * h2.vy = h2.py - h1.py

        // Using Cramer's rule (or direct substitution):
        long det = h1.vx * h2.vy - h1.vy * h2.vx;

        if (det == 0) {
            // Parallel lines
            return false;
        }

        double t1 = (double) (h2.vy * (h2.px - h1.px) - h2.vx * (h2.py - h1.py)) / det;
        double t2 = (double) (h1.vy * (h2.px - h1.px) - h1.vx * (h2.py - h1.py)) / det;

        // Check if intersection is in the future for both hailstones
        if (t1 < 0 || t2 < 0) {
            return false;
        }

        double intersectX = h1.px + t1 * h1.vx;
        double intersectY = h1.py + t1 * h1.vy;

        return intersectX >= min && intersectX <= max && intersectY >= min && intersectY <= max;
    }
}
